text/0000-structural-records.md

```diff
@@ -0,0 +1,1170 @@
+  }
+  ```
+
++ For `PartialEq`, each field is compared with same field in `other: Self`.
+
++ For `ParialOrd` and `Ord`, lexicographic ordering is used based on
+  the name of the fields and not the order given because structural records
+  don't respect the order in which the fields are put when constructing or
+  giving the type of the record.
+
++ For `Debug` the same lexicographic ordering for `Ord` is used.
+  As an example, when printing out `my_record` as with:
+
+  ```rust
+  let my_record = { beta: true, alpha: 42u8, gamma: "Dancing Ferris" };
+  println!("{:#?}", my_record);
+  ```
+  
+  the following would appear:
+  
+  ```rust
+  {
+      alpha: 42,
+      beta: true,
+      gamma: "Dancing Ferris"
+  }
+  ```
+
++ For `Hash`, the same ordering of the fields as before is used and then
+  `self.the_field.hash(state)` is called on each field in that order.
+  For example:
+
+  ```rust
+  self.alpha.hash(state);
+  self.beta.hash(state);
+  self.gamma.hash(state);
+  ```
+
+For auto traits (e.g. `Send`, `Sync`, `Unpin`), if all field types implement
+the auto trait, then the structural record does as well.
+For example, if `A` and `B` are `Send`, then so is `{ x: A, y: B }`.
+
+A structural record is `Sized` if all field types are.
+If the lexicographically last field of a structural record is `?Sized`,
+then so is the structural record. If any other field but the last is
+`?Sized`, then the type of the structural record is not well-formed.
+
+### Implementations and orphans
+
+[RFC 2451]: https://github.com/rust-lang/rfcs/pull/2451
+
+It is possible to define your own implementations for a structural record.
+The orphan rules that apply here are those of [RFC 2451] by viewing a structural
+record as a positional tuple after sorting the elements lexicographically.
+
+For example, if a `trait` is crate-local, we may implement it for a record:
+
+```rust
+trait Foo {}
+
+impl Foo for { alpha: bool, beta: u8 } {}
+```
+
+Under 2451, we can also write:
+
+```rust
+struct Local<T>(T);
+
+impl From<()> for Local<{ alpha: bool, beta: u8 }> { ... }
+```
+
+This is valid because `Local` is considered a local type.
+
+However, a structural record itself isn't a local type, so you cannot write:
+
+```rust
+struct A;
+
+impl From<()> for { alpha: A, beta: u8 } { ... }
+```
+
+This is the case even though `A` is a type local to the crate.
+
+The behaviour for inherent implementations is also akin to tuples.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Grammar
+[grammar]: #grammar
+
+Given:
+
+```rust
+field ::= ident | LIT_INTEGER ;
+```
+
+We extend the expression grammar with:
+
+```rust
+expr ::= ... | expr_srec ;
+
+expr_srec ::= "{" field_init "," field_inits_tail? "}" ;
+field_inits_tail ::= field_inits ","? | (field_inits ",")? DOTDOT expr ;
+field_inits ::= field_init ("," field_init)* ;
+field_init ::= ident | field ":" expr ;
+```
+
+Note that this grammar permits `{ 0: x, 1: y }`.
```

Comment:
🤔 But the following is a syntax error at the moment:

```rust
struct S { 0: u8, 1: u8 }
```

and what happens if you mix numbers and names or doesn't start with 0:

```rust
{ 6: x, what: y }
```

So while this simplifies the grammar I think numbers should still be rejected semantically.

Reply:
> But the following is a syntax error at the moment:

That's true... I've been thinking about it again some more after we closed https://github.com/rust-lang/rfcs/issues/1595 and structural records gives it some more motivation to do that as well.

> and what happens if you mix numbers and names or doesn't start with 0:

At the end of the grammar section I've written:
> Post macro expansion, we check that positional and named fields are not mixed.
We also check that there are no gaps in the positional fields when sorted.

That should deal with it...

> So while this simplifies the grammar I think numbers should still be rejected semantically.

I'll add an unresolved question about this then :)